src/librbd/internal.cc

```diff
@@ -1850,29 +1851,65 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
       }
       assert(m_bl->length() == (size_t)r);
 
-      if (m_bl->is_zero()) {
+      if (!m_sparse_size) {
```

Comment:
Nit: for clarity, when ```m_sparse_size``` is zero, just assume it's set to the object size (```1 << ictx->order```) so that you can use the same code path and logic for both cases.